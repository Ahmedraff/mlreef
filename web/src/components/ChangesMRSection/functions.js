import { Base64 } from 'js-base64';
import Base64ToArrayBuffer from 'base64-arraybuffer';
import { getFileDifferences } from '../../functions/apiCalls';
import CommitsApi from 'apis/CommitsApi';
import { isImageFormat } from 'functions/dataParserHelpers';

const commitsApi = new CommitsApi();

export const fetchFileDiffInfo = (projectId, aheadCommits) => Promise.all(aheadCommits.map(async (commit, index) => {
  const { id: commitId } = commit;
  const commitDiffs = await commitsApi.getCommitDiff(projectId, commitId, 0, false);
  return await Promise.all(commitDiffs?.body?.map(async (commitDiff) => {
    const {
      previousVersionFile,
      nextVersionFile,
      imageFileSize,
    } = await getFileDifferences(projectId, commitDiff, commit.parent_ids[0], commit.id);
    let previousVersionFileParsed = previousVersionFile;
    let nextVersionFileParsed = nextVersionFile;
    let sizeDeleted = 0;
    let sizeAdded = 0;
    const fileName = commitDiff.old_path.split('/').slice(-1)[0];
    if (previousVersionFile && !isImageFormat(fileName)) {
      previousVersionFileParsed = Base64.decode(
        Base64ToArrayBuffer.encode(previousVersionFile),
      );
      sizeDeleted = previousVersionFile.byteLength;
    }
    if (nextVersionFile && !isImageFormat(fileName)) {
      nextVersionFileParsed = Base64.decode(Base64ToArrayBuffer.encode(nextVersionFile));
      sizeAdded = nextVersionFile.byteLength;
    }
    const path = commitDiff.new_path
      ? commitDiff.new_path
      : commitDiff.old_path;
    const fileExtension = path.split('.').pop().toLowerCase();
    return {
      fileName,
      fileExtension,
      previousVersionFileParsed,
      nextVersionFileParsed,
      id: `${commit.id} ${index}`,
      sizeDeleted,
      sizeAdded,
      imageFileSize,
    };
  }));
}));
