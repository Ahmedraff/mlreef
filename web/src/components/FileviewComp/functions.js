import FilesApi from 'apis/FilesApi.ts';
import CommitsApi from 'apis/CommitsApi.ts';

const commitsApi = new CommitsApi();

const filesApi = new FilesApi();

const getCommit = (projectId, lastCommitId) => commitsApi
  .getCommitDetails(projectId, lastCommitId);

const getFileAndInformation = (gid, file, branch, commit) => filesApi
  .getFileData(gid, file?.includes('/') ? encodeURIComponent(file) : file, branch || commit)
  .then(async (fData) => {
    const commitInfoDet = await getCommit(gid, fData.last_commit_id);
    return {
      fData,
      commitInfoDet,
    };
  });

export default {
  getFileAndInformation,
};
