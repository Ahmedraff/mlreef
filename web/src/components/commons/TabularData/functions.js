import genUUID from 'uuid/v1';

const parseNumber = (text) => {
  const num = parseFloat(text);

  return Number.isNaN(num) ? text : num;
};

export const transpose = (matrix) => matrix[0].map((_, ci) => matrix.map((row) => row[ci]));

const sanatizeMatrix = (matrix) => {
  const head = matrix[0];
  const tail = matrix[matrix.length - 1];

  const sameLength = head.length === tail.length;
  const lastRowfilled = tail.every((col) => col !== undefined);

  return sameLength && lastRowfilled ? matrix : matrix.slice(0, -1);
};

export const cvsToArray = (text, separator = ',') => {
  const matrix = text.split('\n')
    .map((line) => line.split(separator)
      .map(parseNumber));

  return sanatizeMatrix(matrix);
};

export const jsonToArray = (input) => {
  const items = Array.isArray(input) ? input : Object.values(input);

  if (!items.length) return items;
  // const row = Object.values(input)[0];
  const row = items[0];
  const header = Object.keys(row).map((label) => label);

  // const data = Object.values(input)
  const data = items
    .map((cols) => Object.values(cols).map((val) => val));

  const matrix = [header].concat(data);

  return sanatizeMatrix(matrix);
};

export const arrayToRichData = (matrix, separator = ',') => {
  const header = matrix[0];

  return {
    id: genUUID(),
    meta: {
      separator,
      timestamp: new Date(),
    },
    matrix,
    // data: flat(matrix.map((row, y) => row.map((value, x) => ({ x, y, value }))))
    data: matrix.map((cols, y) => ({
      id: y,
      cols: cols.map((value, x) => ({
        x,
        y,
        value,
        label: header[x],
      })),
    })),
  };
};
