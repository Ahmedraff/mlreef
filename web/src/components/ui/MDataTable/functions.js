const extractField = (row, x) => row.cols.find((col) => col.x === x)?.value;

export const sortDesc = (x) => (rows) => rows.slice()
  .sort((a, b) => extractField(b, x) - extractField(a, x));

export const sortAsc = (x) => (rows) => rows.slice()
  .sort((a, b) => extractField(a, x) - extractField(b, x));

export const round = (decs) => (input) => {
  const decimals = parseInt(decs, 10);

  if (Number.isNaN(decimals)) return input;

  if (decimals > 12) return input;

  const num = parseFloat(input);
  if (Number.isNaN(num)) return input;

  return Math.round(num * (10 ** decimals)) / (10 ** decimals);
};
