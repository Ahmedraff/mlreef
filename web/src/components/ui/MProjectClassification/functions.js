import { ALGORITHM, DATA_OPERATION, MODEL, OPERATION, VISUALIZATION } from "dataTypes";

export const getSearchable = (feSearchableTypes) => {
  switch (feSearchableTypes) {
    case DATA_OPERATION:
      return OPERATION;
    case MODEL:
      return ALGORITHM;
    default:
      return VISUALIZATION;
  }
}