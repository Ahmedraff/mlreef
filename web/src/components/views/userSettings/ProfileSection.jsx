import React, { useState, useEffect, useRef } from 'react';
import { connect } from 'react-redux';
import { toastr } from 'react-redux-toastr';
import MButton from 'components/ui/MButton';
import PropTypes from 'prop-types';
import Base64ToArrayBuffer from 'base64-arraybuffer';
import UserApi from 'apis/UserApi';
import { INFORMATION_UNITS } from 'domain/informationUnits';
import './Profile.scss';

const userApi = new UserApi();

const ProfileSection = (props) => {
  const { user: { auth, username, gitlab_id: gitlabId, userInfo: { avatar_url: avatarUrl } } } = props;
  const fileMaxSize = 200;
  const imageInput = useRef();
  const [imgBase, setImgBase] = useState(avatarUrl);
  const [avatarFile, setAvatarFile] = useState(null);
  const [file, setImageName] = useState('No file chosen');
  const [userName, setUserName] = useState(username);
  const [status, setUserStatus] = useState('');
  const [uploading, setUploading] = useState(false);

  useEffect(() => {
    userApi.getUserStatus(gitlabId)
      .then((res) => res.message !== null ? setUserStatus(res.message) : setUserStatus(''))
      .catch(() => toastr.error('Error', 'Could not fetch the status'));
  }, [gitlabId]);

  const profileStatusUpdate = () => userApi.updateUserStatus(status)
    .then(() => {
      setUploading(false);
      toastr.success('Success', 'Status has been updated successfully');
    })
    .catch(() => toastr.error('Error', 'Status cannot be updated'));

  const updateProfileSettings = (e) => {
    setUploading(true);
    e.preventDefault();
    const payload = {
      admin: auth,
      gitlabId,
      name: userName,
    };
    if (avatarFile && typeof avatarFile !== 'undefined') {
      const reader = avatarFile.stream().getReader();
      reader.read().then((content) => {
        payload.avatar = Base64ToArrayBuffer.encode(content.value);
        profileStatusUpdate(payload);
      });
    } else {
      profileStatusUpdate(payload);
    }
  };

  const handleImageSelect = (event) => {
    const selectedFiles = event.target.files[0];
    if (selectedFiles !== undefined) {
      if ((selectedFiles.size / INFORMATION_UNITS.KILOBYTE) > fileMaxSize) {
        toastr.error('Error', 'File size exceeds 200Kb');
      } else {
        setAvatarFile(selectedFiles);
        setImageName(selectedFiles.name);
        const reader = new FileReader();
        reader.addEventListener('load', () => {
          setImgBase(reader.result);
        }, false);
        reader.readAsDataURL(selectedFiles);
      }
    }
  };

  return (
    <div className="user-form">
      <div className="row d-flex">
        <div className="pl-4 pr-4 left-row">
          <h4 className="mt-0 display-1">Public Avatar</h4>
          <p>
            You can change your avatar here or remove the
            current avatar to revert to
            <a href="https://gravatar.com" target="_blank" rel="noopener noreferrer"> gravatar.com</a>
          </p>
        </div>
        <div className="pl-4 right-row">
          <div className="mb-0" style={{ float: 'left' }}>
            <a target="_blank" href={avatarUrl} rel="noopener noreferrer">
              <img style={{ borderRadius: '50%' }} className="mr-3 avatar resposniveAvatar" width="140" height="140" src={imgBase} alt="" />
            </a>
          </div>
          <h5 className="mt-0">Upload new avatar</h5>
          <div className="mt-1 mb-2">
            <button
              type="button"
              className="btn btn-basic-dark"
              onClickHandler={() => imageInput.current.click()}
            >
              Choose File
            </button>
            <span className="ml-2" style={{ position: 'relative', display: 'inline-block' }}>{file}</span>
            <input
              ref={imageInput}
              id="image-file"
              className="d-none invisible"
              type="file"
              accept="image/*"
              onChange={handleImageSelect}
            />
          </div>
          <div className="mb-0 mt-3 t-secondary">The maximum file size allowed is 200Kb</div>
        </div>
      </div>
      <hr />
      <div className="row d-flex mt-4">
        <div className="pl-4 pr-4 left-row">
          <h4 className="mt-0">Current Status</h4>
          <p>
            This emoji and message will appear on you profile
            and throughout the interface.
          </p>
        </div>
        <div className="pl-4 pr-4 right-row">
          <label htmlFor="user_status"><h5 className="m-0">Your Status</h5></label>
          <input
            className="mt-3 w-auto"
            id="user_status"
            type="text"
            value={status === null ? '' : status}
            placeholder="What's your status?"
            onChange={(e) => setUserStatus(e.target.value)}
          />
        </div>
      </div>
      <hr />
      <div className="row d-flex">
        <div className="pl-4 pr-4 mt-4 left-row">
          <h4 className="mt-0">Main Settings</h4>
          <p>
            This information will appear on your
            profile.
          </p>
        </div>
        <div className="pr-4 d-flex right-row">
          <div className="pr-4 pl-4 mb-4 left-col">
            <label htmlFor="user_name"><h5>Full name</h5></label>
            <input
              className="mt-1 w-100"
              id="user_name"
              type="text"
              placeholder="Current name"
              value={userName}
              onChange={(e) => setUserName(e.target.value)}
            />
            <p hidden>Please enter your name</p>
            <p className="mt-2" style={{ color: '#919191' }}>Enter your name, so people you know can recognize you.</p>
          </div>
          <div className="pr-4 pl-4 mb-4 right-col">
            <label htmlFor="user_id"><h5>User ID</h5></label>
            <input
              readOnly
              className="mt-1 w-50 muted bg-light"
              id="user_id"
              type="text"
              value={gitlabId}
            />
          </div>
        </div>
      </div>
      <div className="pl-4 mt-4">
        <MButton
          type="button"
          waiting={uploading}
          className="btn btn-primary"
          onClick={(e) => updateProfileSettings(e)}
        >
          Update Profile Settings
        </MButton>
      </div>
    </div>
  );
};

ProfileSection.propTypes = {
  user: PropTypes.shape({
    auth: PropTypes.bool.isRequired,
    username: PropTypes.string.isRequired,
    gitlab_id: PropTypes.number.isRequired,
    userInfo: PropTypes.shape({
      avatar_url: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
};

function mapStateToProps(state) {
  return {
    user: state.user,
  };
}

export default connect(mapStateToProps)(ProfileSection);
