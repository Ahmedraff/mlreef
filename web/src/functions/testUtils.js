import { createStore, applyMiddleware } from 'redux';

import rootReducer from '../reducers/index';
import { middlewares } from '../configureStore';

// eslint-disable-next-line import/prefer-default-export
export const storeFactory = (initialState) => {
  const createStoreWithMiddleware = applyMiddleware(...middlewares)(createStore);
  return createStoreWithMiddleware(rootReducer, initialState);
};
