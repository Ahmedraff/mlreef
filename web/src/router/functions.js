/**
 * Compose like: f.g x = g(f(x))
 *
 * @example compose (fn4, fn3, fn2, fn1) (arg) === fn4(fn3(fn2(fn1(arg))))
 *
 * @param {Array[Function]} fns sequence of functions.
 * @param arg input parameter.
 *
 * @return far left funcion return.
 */
export const c = (...fns) => (arg) => fns.reverse().reduce((acc, fn) => fn(acc), arg);

/**
 * Return the recorded route that matches name.
 */
export const findRouteInfo = (routes) => ({ name }) => routes.find((route) => route.name === name);

/**
 * Replace placeholders by parameters. e.g. /:user/messages -> /nick/messages.
 *
 * @param {Object(Route)} to implicit.
 * @param {Object} params e.g. key-val object, from the to object.
 * @param {Object(Route)} record the route from routes.
 *
 * @return {Object(Route)} route with path modified.
 */
export const replaceParams = (to) => (record = {}) => {
  if (to.params) {
    const path = Object.entries(to.params)
      .reduce((final, [k, v]) => final.replace(new RegExp(`:${k}`), v), record.path || '');

    return { ...record, path };
  }

  return record;
};

/**
 * Add the hash part. e.g. #last-menu-item
 *
 * @param {Object(Route)} to implicit.
 * @param {String} hash e.g. last-menu-item, from the to object.
 * @param {Object(Route)} record the route from routes.
 *
 * @return {Object(Route)} route with path modified.
 */
export const addHash = (to) => (record) => !to.hash ? record
  : { ...record, path: `${record.path}#${to.hash}` };

export const makeQuery = (query) => Object.entries(query)
  .map(([k, v]) => `${k}=${v}`)
  .join('&');

/**
 * Add the query search. e.g. ?user=nickname&priority=high
 *
 * @param {Object(Route)} to implicit.
 * @param {Object} query e.g. key-val object, from the to object.
 * @param {Object(Route)} record the route from routes.
 *
 * @return {Object(Route)} route with path modified.
 */
export const addQuery = (to) => (record) => !to.query ? record
  : { ...record, path: `${record.path}?${makeQuery(to.query)}` };

export const unpack = ({ path }) => path;

export const resolveRoute = (routes, to) => c(
  unpack, addQuery(to), addHash(to), replaceParams(to), findRouteInfo(routes),
)(to);
