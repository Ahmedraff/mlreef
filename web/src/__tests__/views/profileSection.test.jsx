import React from 'react';
import { shallow } from 'enzyme';
import { storeFactory } from 'functions/testUtils';
import { userProfileMock } from 'testData';
import ProfileSection from 'components/views/userSettings/ProfileSection';

const user = {
  auth: true,
  userInfo: {
    avatarUrl: 'https://www.gravatar.com/avatar/d64636c9c4cf15dd5c9e1ed6ab529100?s=80&d=identicon',
    id: 2,
    name: 'mlreef',
  },
};

const setup = () => {
  const store = storeFactory({
    user: userProfileMock,
  });
  const wrapper = shallow(
    <ProfileSection store={store} user={user} />,
  );
  const afterDive = wrapper.dive().dive();
  return afterDive;
};

describe('UI renders user profile with user data', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = setup();
  });

  test('assert that user profile is rendered correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('assert that file input is present and has the right props', () => {
    const fileInput = wrapper.find('[type="file"]');
    expect(fileInput).toHaveLength(1);
    const fileInputProps = fileInput.first().props();
    expect(fileInputProps.accept).toBe('image/*');
  });

  test('assert that user avatar is displayed', () => {
    expect(wrapper.find('img').props().src).toEqual(user.userInfo.avatarUrl);
  });
});
