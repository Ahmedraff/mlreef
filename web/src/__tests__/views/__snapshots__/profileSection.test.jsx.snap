// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI renders user profile with user data assert that user profile is rendered correctly 1`] = `ShallowWrapper {}`;
